File: VB/DocumentExportWithTokenBasedAuthentication/Controllers/ReportViewerController.vb
Imports DevExpress.Web.Mvc.Controllers
Imports DevExpress.XtraReports.Web.WebDocumentViewer
Imports T680906.Services
Imports System.Web.Mvc

Namespace T680906.Controllers
	Public Class ReportViewerController
		Inherits WebDocumentViewerApiController

		Private documentExportService As IExportResultProvider
		Public Sub New()
			documentExportService = DirectCast(DefaultWebDocumentViewerContainer.Current.GetService(GetType(IExportResultProvider)), IExportResultProvider)
		End Sub

		<Authorize>
		Public Overrides Function Invoke() As ActionResult
			Return MyBase.Invoke()
		End Function

		<HttpGet>
		Public Function GetExportResult(ByVal token As String, ByVal fileName As String) As ActionResult
			Dim exportResult As ExportResult = Nothing
			If Not documentExportService.TryGetExportResult(token, exportResult) Then
				Return New HttpNotFoundResult("Exported document was not found. Try to export the document once again.")
			End If
			Dim fileResult = File(exportResult.GetBytes(), exportResult.ContentType)
			If exportResult.ContentDisposition <> System.Net.Mime.DispositionTypeNames.Inline Then
				fileResult.FileDownloadName = exportResult.FileName
			End If

			Return fileResult
		End Function
	End Class
End Namespace

File: CS/DocumentExportWithTokenBasedAuthentication/Controllers/ReportViewerController.cs
using DevExpress.Web.Mvc.Controllers;
using DevExpress.XtraReports.Web.WebDocumentViewer;
using T680906.Services;
using System.Web.Mvc;

namespace T680906.Controllers
{
    public class ReportViewerController : WebDocumentViewerApiControllerBase
    {
        IExportResultProvider documentExportService;
        public ReportViewerController() {
            documentExportService = (IExportResultProvider)DefaultWebDocumentViewerContainer.Current.GetService(typeof(IExportResultProvider));
        }

        [Authorize]
        public override ActionResult Invoke() {
            return base.Invoke();
        }

        [HttpGet]
        public ActionResult GetExportResult(string token, string fileName) {
            ExportResult exportResult;
            if (!documentExportService.TryGetExportResult(token, out exportResult)) {
                return new HttpNotFoundResult("Exported document was not found. Try to export the document once again.");
            }
            var fileResult = File(exportResult.GetBytes(), exportResult.ContentType);
            if (exportResult.ContentDisposition != System.Net.Mime.DispositionTypeNames.Inline) {
                fileResult.FileDownloadName = exportResult.FileName;
            }

            return fileResult;
        }
    }
}